Wikipedia/Code/MWKLanguageLinkController.m

```diff
@@ -104,12 +125,10 @@ - (nullable MWKLanguageLink *)appLanguage {
     NSArray *preferredLanguageCodes = [self readPreferredLanguageCodes];
     return [preferredLanguageCodes wmf_mapAndRejectNil:^id(NSString *langString) {
         return [self.allLanguages wmf_match:^BOOL(MWKLanguageLink *langLink) {
-            //Note, sometimes the device iOS language codes will return a code that doesn't line up with the Wikipedia language codes we have set,
-            //so we are also checking for a match against the altISOCode (currently only set for "no.wikipedia.org")
-            //Fixes https://phabricator.wikimedia.org/T276645
-            return [langLink.contentLanguageCode isEqualToString:langString] ||
-                   (langLink.altISOCode &&
-                    [langLink.altISOCode isEqualToString:langString]);
+            // The Norwegian language code 'nb' used by the OS should be automatically converted to 'no' used by Wikipedia during onboarding.
+            // Issue described in  https://phabricator.wikimedia.org/T276645
+            NSAssert(![langString isEqualToString:@"nb"], @"Inaccurate language code found in preferred language codes. Language code 'nb' should be 'no'.");
```

Comment:
I came across a situation where this assertion is hit - 

1. On a device with Norwegian Bokmal as a language, check out releases/6.7.4 and run.
2. Go through onboarding and add any language on the choose languages screen. (It will be in that weird empty state due to "nb" not being found on the old version).
3. Check out this branch and run.
4. Assertion is hit

It seems on `MWKDataStore`'s init, this code is run before migration happens.

Reply:
Looking at it, the preferred site URLs, which are based on the preferred languages are set on the explore feed *before* migration happens.

I need to investigate a bit more but there are a few things I want to check:

- In 6.7.4 Norwegian in the preferred languages list is just ignored, so won't appear in the explore feed.
- In 6.8 it is not ignored because of the altISOCode check
- In 6.8.1 upgrading from 6.7.4, it would be ignored when preferredSiteURLs is initially set. What happens to the explore feed settings with Norwegian going from 6.7.4 to 6.8 / 6.8.1?

2. It seems (on 6.7.4 at least) that if you go through onboarding without making a change to preferred languages that preferred languages are never set.

It seems like that means that every time the app is launched, we end up getting the same list we would get from onboarding. That doesn't seem like what we'd want to do though.

Reply:
I've updated my approach for `-preferredLanguages` in the most recent commit.

The new approach does the following:
- Caches the preferred languages in a `cachedPreferredLanguages` array that is invalidated when preferred languages change. This prevents repeated fetching and searching to generate the same array repeatedly.
- Uses the new `+languageCodeForISOLanguageCode:` method added in an earlier commit for article language links. This will do the altISOCode -> languageCode substitution even before the language variant migration updates the settings.